### ISSUE TITLE
Transformations breaking with custom columnNames

### ISSUE THREAD

**Issue by toddgeist** [balderdashy/waterline#1072]
I am having difficulty getting associations to work with `columnName` attributes.  I think the columnName transformation is missing the fact that this is a foreignKey and blowing up the related record.

Shouldn't this line in https://github.com/balderdashy/waterline/blob/master/lib/waterline/core/transformations.js#L61
`if(attr === attributes[attr][key]) return;`

really be
 `if(attr === attributes[attr][key] || attributes[attr]['foreignKey']  ) return;`

???

Otherwise the entire related model is overwritten with just the value of it's primary key.  When I make this change it works and I get the expected result. ( also see model below )

```json
{
  "destinationLocation": {
    "_id": "6152",
    "description": "xxxx"
  },
  "_id": "132086726030911437678303430040771397634",
  "date": "06/08/2015",
  "_id_action": "13",
  "_id_company_Destination": "320",
  "_id_company_Source": "66",
  "_id_driver": "",
  "_id_job": "334228432983455469160919776409968341717",
  "_id_substance": "4",
  "_id_location_Destination": "6152",
  "modid": "12",
  "recid": "34"
}

```
If I don't make that change I get this 

```json
{
  "destinationLocation": {
    "0": "6",
    "1": "1",
    "2": "5",
    "3": "2",
    "bold": "\u001b[1m6152\u001b[22m",
    "underline": "\u001b[4m6152\u001b[24m",
    "strikethrough": "\u001b[9m6152\u001b[29m",
    "italic": "\u001b[3m6152\u001b[23m",
    "inverse": "\u001b[7m6152\u001b[27m",
    "grey": "\u001b[90m6152\u001b[39m",
    "black": "\u001b[30m6152\u001b[39m",
    "yellow": "\u001b[33m6152\u001b[39m",
    "red": "\u001b[31m6152\u001b[39m",
    "green": "\u001b[32m6152\u001b[39m",
    "blue": "\u001b[34m6152\u001b[39m",
    "white": "\u001b[37m6152\u001b[39m",
    "cyan": "\u001b[36m6152\u001b[39m",
    "magenta": "\u001b[35m6152\u001b[39m",
    "greyBG": "\u001b[49;5;8m6152\u001b[49m",
    "blackBG": "\u001b[40m6152\u001b[49m",
    "yellowBG": "\u001b[43m6152\u001b[49m",
    "redBG": "\u001b[41m6152\u001b[49m",
    "greenBG": "\u001b[42m6152\u001b[49m",
    "blueBG": "\u001b[44m6152\u001b[49m",
    "whiteBG": "\u001b[47m6152\u001b[49m",
    "cyanBG": "\u001b[46m6152\u001b[49m",
    "magentaBG": "\u001b[45m6152\u001b[49m",
    "rainbow": "\u001b[31m6\u001b[39m\u001b[33m1\u001b[39m\u001b[32m5\u001b[39m\u001b[34m2\u001b[39m",
    "zebra": "6\u001b[7m1\u001b[27m5\u001b[7m2\u001b[27m",
    "stripColors": "6152",
    "zalgo": "6̷͇͎̦̱̹̜̘͚̣̼͇̜͇̮̱̱̦̦̳͈̟͚̼̯͉̠͖̲̝͚͖̜͇͖̮̹͉̟̳͔͉̫͉̬̬̖̥̬̜̻̖̮̹̙̲̤̥͍̲̪̍͛̇͐ͮ͗̍͌̓ͯͭ͂͗͋͒ͬ͒̚1̬̠͕̘͔̟̜͈̰̦͇̬͎͇̼̞̩̱̬̲̲̦̝͍̥̞̯̱̱̬͙̬͚͚̮͙̘̮̟͚̠͓̟͎̹̲̝͖̋͂̌̎͗ͩͨͮͥ̊ͣ͑̽̑ͣ͋̚͠ͅ5̴̻͇̪̮̝̝͖̱̻̭̯̗͇̣͚̗̥̲̝̱̣̹̮̰͎̠̘̬̲̭̗̲̺͔̭̫̬̥̞͚̹̯̳̗̖̙̩͈̘̜̖̲͚̘̥̮͙̲̤͇͈̹̰̼ͤ̎͌ͥ͑ͫͣͭ̀ͬ̉̊ͪ̚ͅ2̨̳͇͇̦͎̲̞̼̱̣͉̱̣͎̠͍̞̬̪̟̦̤̣̹̗̙͈͇̙̰̠͔͔͚͙̱̙̱̝̰̦͈̻̠̺̥͍̟̆̿ͨͫͨͪ͆̾̊̒ͦͦ̋ͭ͒̅͋̑̚ͅͅͅ",
    "silly": "\u001b[31m6\u001b[39m\u001b[33m1\u001b[39m\u001b[32m5\u001b[39m\u001b[34m2\u001b[39m",
    "input": "\u001b[30m6152\u001b[39m",
    "verbose": "\u001b[36m6152\u001b[39m",
    "prompt": "\u001b[90m6152\u001b[39m",
    "info": "\u001b[32m6152\u001b[39m",
    "blank": "\u001b[37m6152\u001b[39m",
    "data": "\u001b[90m6152\u001b[39m",
    "help": "\u001b[36m6152\u001b[39m",
    "warn": "\u001b[33m6152\u001b[39m",
    "debug": "\u001b[34m6152\u001b[39m",
    "error": "\u001b[31m6152\u001b[39m",
    "crit": "\u001b[31m6152\u001b[39m"
  },
  "_id": "132086726030911437678303430040771397634",
  "date": "06/08/2015",
  "_id_action": "13",
  "_id_company_Destination": "320",
  "_id_company_Source": "66",
  "_id_driver": "",
  "_id_job": "334228432983455469160919776409968341717",
  "_id_substance": "4",
  "modid": "12",
  "recid": "34"
}
```
Which is clearly wrong.  The first four items "6" , "1", "5", "2"  are the value of the primary key 6152.

Model
=========

```javascript
module.exports = {
 
  // FileMaker will handle these for us
  autoPK : false,
  autoCreatedAt: false,
  autoUpdatedAt: false,
  tableName : 'apiTickets',
 
  attributes: {
 
    //setup the primary key
    _id: {
      type: 'string',
      primaryKey: true,
      unique: true
    },
 
    // sets up an association to locations
    location : {
      model : 'location',
      columnName : '_id_location_Destination'
    },
 
  }
 
}
```

**Comment by devinivy** [comment]
I think you're right.  Oy!  Very odd that we haven't run into this.  Let's work on writing a failing test to fix.  I'm not _convinced_ for sure that it's as simple as the fix you've suggested, but I'm honestly unsure.  We'll have to look at when the transformation is expected to occur in the process of tacking joined records onto the model.

Do you experience this during a `Collection.find`?  Do you know why your string class has all those properties for display use in CLI?  Also, do you mind letting us know your sails version?  Lastly, does your model use the attribute `destinationLocation` or `location`?

**Comment by toddgeist** [comment]
yes, sorry I changed the name of the attribute a few times. it should be

```javascript
module.exports = {

  // FileMaker will handle these for us
  autoPK : false,
  autoCreatedAt: false,
  autoUpdatedAt: false,
  tableName : 'apiTickets',

  attributes: {

    //setup the primary key
    _id: {
      type: 'string',
      primaryKey: true,
      unique: true
    },

    // ** CORRECRECTED **
    destinationLocation : {
      model : 'location',
      columnName : '_id_location_Destination'
    },

  }

}
```

**Comment by toddgeist** [comment]
I am using sails, version "0.11.1", and waterline is "0.10.19"

**Comment by toddgeist** [comment]
I am getting there through finders.findOne()

https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/finders/basic.js#L28

But more specifically it shows up in the `returnResult` function
https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/finders/basic.js#L150-L158

**Comment by toddgeist** [comment]
as for all those props coming from the command line output... I am not sure where that comes from.  but I can tell you that if you examine the result just after the transformations occur you'll see 
`destinatationLocation : 6152` 
not 
`destinatationLocation : {...} `  where {...} is the correct associated object.

So what it looks like to me is that when record is re-serialized, the `6152` gets treated like an Array that is then turned into an Object.  Somewhere along the line it picks up all those extra properties.

**Comment by toddgeist** [comment]
I am happy to work on writing a failing test. But I'll need some guidance. I know my way around a little having just published [sails-filemaker](https://www.npmjs.com/package/sails-filemaker).  But I am not sure where to write this test.

here perhaps? https://github.com/balderdashy/waterline/blob/master/test/unit/query/query.findOne.js

and won't we need additional fixtures, to load associations with `columnName` ?

**Comment by particlebanana** [comment]
Is this different than https://github.com/balderdashy/waterline/pull/1054

**Comment by toddgeist** [comment]
I believe it is different.  This is doing a transform on an attribute with a columnName prop. The problem is that the attribute is an associated collection with a columnName prop to indicate a non standard foreign Key field.

https://github.com/balderdashy/waterline/blob/master/lib/waterline/core/transformations.js#L60-L63

```javascript
      // Set transformation attr to new key
      if(key === 'columnName') {
        if(attr === attributes[attr][key]) return;
        self._transformations[attr] = attributes[attr][key];
      }
```
So it blows out the associated collection with the value of its primary key.  Happy to write a test if somebody can help guide me.

**Comment by dmarcelino** [comment]
@toddgeist, have you taken a look at [Collection.transformer.js](https://github.com/balderdashy/waterline/blob/master/test/integration/Collection.transformer.js) and [association.getter.js](https://github.com/balderdashy/waterline/blob/master/test/integration/model/association.getter.js) tests? Those are probably a good starting point to create a test for this. Let us know if you have more specific questions.

**Comment by toddgeist** [comment]
Ok, that helps.  I'll take a look at it this weekend

thanks

Todd

**Comment by sailsbot** [comment]
Thanks for posting, @toddgeist.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
